### PR TITLE
Fix image sources being clipped at -180 and 180 longitude when terrain is enabled

### PR DESCRIPTION
**Pull request by pstaszek**
Fixed issue #4088.

### TL;DR
Lack for wrapping support for terrain render to texture tiles.

### Lond description.

Terrain tiles are rendered to the texture, which then is rendered to the tile (images as well) so content can't overlap between tiles. There was no support for wrapping, so for every wrap image was rendered on the same position relative to the tile.

See the diagram:
wrapping_issue_illustration

The image on tile with wrap = 0 is rendered at desired position, however on tile with wrap = 1 it is rendered on the same position in relation to a tile as for previous one. It should be actually rendered with shift.

### Before:
before

### After:
after

I also added render test.

## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [ ] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

File: src/source/terrain_source_cache.ts [hunk 1]
```diff
@@ -294,10 +297,14 @@ export class TerrainSourceCache extends Evented {
...
         canonicalTileRanges: {[zoom: string]: CanonicalTileRange}
     ): boolean {
-        return canonicalTileRanges[tileID.canonical.z] &&
-            tileID.canonical.x >= canonicalTileRanges[tileID.canonical.z].minTileX &&
-            tileID.canonical.x <= canonicalTileRanges[tileID.canonical.z].maxTileX &&
-            tileID.canonical.y >= canonicalTileRanges[tileID.canonical.z].minTileY &&
-            tileID.canonical.y <= canonicalTileRanges[tileID.canonical.z].maxTileY;
+        const range = canonicalTileRanges[tileID.canonical.z];
+
+        return range && (
```
Comment (HarelM) [inline src/source/terrain_source_cache.ts]: This will return `null` in case range is `null` instead of `false`.

**Comment by HarelM** [comment]
Can you attach a screen shot of how the render test would look without the fix to make sure the render test covers this scenario?

**Comment by HarelM** [comment]
Is the code coverage report correct by highlighting that the changed code is not covered in tests?
image

**Comment by pstaszek** [comment]
Render test screenshots
On `main` branch:
actual

On current branch (correct)
expected

File: src/source/terrain_source_cache.ts (hunk 1 above)
Comment (pstaszek) [reply src/source/terrain_source_cache.ts]: Nice catch! Fixed.

**Comment by pstaszek** [comment]
For the coverage report, this is a bit more complex.

I assume the coverage report is based on unit tests. Unit tests check only logic responsible for adding and omitting tiles. They do not test matrix computation. This is because I believe tests should be easy to verify, it should be clear if they are correct, while in such tests we would compare some magic numbers in 3D matrices.

This logic is however tested in render test - but only a part of it. For the matrix there are 3 different scenarios, depending on a image and terrain tile zoom level (might be higher, lower or the same). I tested only the case which was relevant for the reported issue. My assumption was that I shouldn't add too many render tests, I didn't want to make them too granular covering very narrow edge cases.

Let me know if and how I can improve that. Should I add unit or render tests? Or current state is ok?

**Comment by HarelM** [comment]
Both render tests and unit tests are added to the coverage report, I think you might be able to find it in codecov site, unfortunately it doesn't get printed here due to a bug on their side. 
The requirement in general is that new code that was added will be covered (hopefully fully) with tests.

**Comment by HarelM** [comment]
As far as I can tell from the code coverage report, the last else was not covered before this PR and seems like it is still not covered after this PR.
I'm not sure I know how to get to this code path, but since you have changed it, I think it would be wise to try and get to it using either a render test or unit test.
https://github.com/pstaszek/maplibre-gl-js/blob/4884e54f27a509b3b6535f0b7623fee6023d586e/src/source/terrain_source_cache.ts#L239

Let me know if there's anything else I can help with.

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,7 @@
 
 - Fix setting visibility on custom layer ([#6883](https://github.com/maplibre/maplibre-gl-js/issues/6883) (by [melitele](https://github.com/melitele))
 - Hide leading and trailing control characters in `format` expressions ([#6907](https://github.com/maplibre/maplibre-gl-js/pull/6907) by [1ec5](https://github.com/1ec5))
+- Fix image sources being clipped at -180 and 180 longitude when terrain is enabled ([#4088](https://github.com/maplibre/maplibre-gl-js/issues/4088))
```
Comment (HarelM) [inline CHANGELOG.md]: Add "by" link please

**Comment by HarelM** [comment]
Looks good overall, does this fix needed for canvas source as well?

**Comment by pstaszek** [comment]
I checked canvas source as well - it is also fixed. `CanvasSource` inherits most of the logic from `ImageSource`.
I also added "by" link.
